spec/models/address_spec.rb

```diff
@@ -10,8 +10,8 @@
   end
 
   describe ".has_street_address?" do
-    let!(:address_without_street_address) { create(:address, :no_street_address) }
-    let!(:address_with_street_address) { create(:address) }
+    let(:address_without_street_address) { create(:address, :no_street_address) }
+    let(:address_with_street_address) { create(:address) }
```

Comment:
> Do we have a preference between `let` and `let!`? 

We do, its avoiding `let` or `let!` as much as possible, with `let!` being the worse of the two. 😁 Both of them have very real problems, with `let!`  being more severe. I've been trying to work to reduce our reliance on both as I can.

I added the [lets_not rachet](https://github.com/simpledotorg/simple-server/blob/5a21196f7db598389b12e539d27ca63761c163f3/spec/rachets/lets_not_spec.rb) as a first step to start to improve our test suite around this. That spec has links to Thoughtbot blog posts around why let is harmful. Their posts are about both `let` and `let!`, but I figured it would be good start with the biggest offender I've seen, which is the bang version.

I realize that there are some valid use cases for `let` with the amount of test data we need to setup. With `let!` though, you end up with magical creation at a distance for every spec, with dependency chains that are really difficult to detangle. The bang version also results in incurring excessive fixture overhead for every spec, which is important for a test suite as large as ours. Consider this contrived example, which is not too far from many specs we have in our suite:

```ruby
let!(:organization, name: "my-org")
let!(:facility, :facility_group: facility_group)
let!(:facility_group, organization: organization, user: admin)
let!(:patient_1, registration_facility: facility)
let!(:blood_pressure, :controlled, patient: patient_1)
let!(:admin, :admin)

context "patient BP control status" do
  before do
    facility.do_some_other_setup_thing
  end
  
  it "is controlled if recent BP is controlled" do
    expect(patient_1).to be_controlled
  end

  it "has no uncontrolled BPs" do
    expect(patient_1.blood_pressures.uncontrolled).to be_empty
 end

  # Imagine some other developer adds the two specs below after seeing that the test
  # fixtures above didn't really work, but also not wanting to refactor everything for their 
  # particular test.  This happens commonly with tests when their are many lines of
  # `let` / `let!` setup.
  it "is uncontrolled if recent BP is uncontrolled" do
    patient = create(:patient)
    blood_pressure = create(:blood_pressure, :hypertensive, patient: patient)
    expect(patient).to be_uncontrolled
  end

  it "has no BPs if patient has not had a BP recorded" do
    patient = create(:patient)
    expect(patient.blood_pressures).to be_empty
  end
```

Note the last two specs are creating their own fixtures that are completely separate from the `let!` created fixtures. Yes, they could be in their own context block, or they could be extracted to their own `lets`, but the core issues still exist -- the data under test is too far from the test code itself, and we end up creating we more test data then we need. These problems become even worse when you consider the case of nested contexts and nested `before` blocks.
In a real world test you may have to scroll 100 lines between the rows of setup and the particular spec...that doesn't make for a maintainable or useful spec long-term.

Here is an alternative with many less `lets`, with a single judicious use of a `let` that is not central to the test:

```
  let(:user) { create(:user, name: "some-test-user) }

  it "is controlled if recent BP is controlled" do
    patient = create(:patient, user: user)
    bp = create(:blood_pressure, :controlled, user: user)
     expect(patient).to be_controlled
  end

  it "has no uncontrolled BPs" do
    patient = create(:patient, user: user)
    bp = create(:blood_pressure, :controlled, user: user)
    expect(patient.blood_pressures.uncontrolled).to be_empty
 end

  it "is uncontrolled if recent BP is uncontrolled" do
    patient = create(:patient, user: user)
    bp = create(:blood_pressure, :controlled, user: user)
    expect(patient).to be_uncontrolled
  end

  it "has no BPs if patient has not had a BP recorded" do
    patient = create(:patient, user: user)
    expect(patient.blood_pressures).to be_empty
  end
```

Here the test setup is right by the assertion, so every spec is easy to read and follow. Data that is not core to the test is in a lazy `let` and used to reduce excessive object creation. its easy to add new specs with particular test data that is needed, and you don't need to read a ton of setup to try to decide if you should build on that or do your own thing.

I realize this is an overly simplified example, but we can use plain old ruby when we need more complex setup. Or we can look into why fixture data is so cumbersome, and try to tackle that problem. The blogposts referenced above go into detail about better approaches.

I hope that is helpful, lemme know if you wanna discuss in more detail on zoom or slack.

Reply:
That makes sense, and I'm onboard with "lets not" (`let` and `let!` both). I think as we reduce the ratchet number though, moving from `let!` to `let` does not offer a significant advantage. I don't have a strong opinion between between the two, except for the few places where a `let` can be confusing that I've mentioned.

Thanks for the detailed examples, super helpful :D